### ISSUE TITLE
Redis performance

### ISSUE THREAD

**Issue by angel0d** [omega8cc/boa#1003]
hi,

we have recently migrated a heavy drupal site from VM on the 2.4 series to a VM on BOA 3.1

i am attaching a graph showing the average response time of redis jumping from less than 1ms to several ms

the VM is on linode and we have already migrated to different hosts to rule out any possible underlying problems

have anyone noticed some thing similar after upgrading to 3.x ?

![redis-25th](https://cloud.githubusercontent.com/assets/10475592/16357746/58d3afb6-3aff-11e6-856c-2703b0620b04.png)

**Comment by pricejn2** [comment]
I've been encountering some Redis performance issues as well, also on a heavily trafficked Drupal site. 

You might try `redis_flush_forced_mode = FALSE` in you site's `boa_site_control.ini`.

**Comment by pricejn2** [comment]
I'd also note, that I see large redis latency spikes when cron is running -- particularly as the memory used by redis increases throughout the day. Do any of those peaks in your chart correspond to cron runs?

**Comment by pricejn2** [comment]
@omega8cc Should this be re-enabled -- https://github.com/omega8cc/boa/blob/master/aegir/conf/global.inc#L275? It looks like the underlying issue has since been closed.

**Comment by pricejn2** [comment]
I've been testing this quite a bit today, and for a large high traffic site using the metatag module. I think it really makes sense to exclude metatag from Redis via `redis_exclude_bins = "cache_metatag"` in your site's `boa_site_control.ini`  (@omega8cc perhaps that would be a good addition for the `/root/.high_traffic.cnf` system control file). Concerns with excessive growth of this cache bin have been expressed in a few d.o. issues, including https://www.drupal.org/node/2145853

Check out a few entries in your Redis `SLOWLOG`:
```
_redis_pass=$(cat /root/.redis.pass.txt)
redis-cli -a $_redis_pass
127.0.0.1:6379> slowlog get 2
```
In my case, they were all related to the `cache_metatag` bin.

Not sure it that fully explains the differences you're seeing between 2.x and 3.x BOA, but in is could be part of the problem.

**Comment by angel0d** [comment]
@pricejn2 thanks a lot for the valuable feedback

I'll be working the next days on the issue

I'll post back what (didn't) worked for us

**Comment by omega8cc** [comment]
We have tried `$conf['redis_eval_enabled'] = TRUE` and it caused serious issues for some sites, so I don't think it is truly fixed. That said, we have not seen any performance problems like reported in this issue. Could be codebase specific, perhaps?

**Comment by pricejn2** [comment]
I have a sneaking suspicion that something is off with the Redis timer for `hmget`, `hget`, and `get` commands in New Relic.

**Comment by pricejn2** [comment]
@angel0d any update on testing progress? 

From my experience, this was 100% site specific and unrelated to BOA 3.1

**Comment by omega8cc** [comment]
By the way, we have added Redis support for Drupal 8 in BOA head. Please test and let us know if you will experience any issues.

![screenshot 2016-07-25 01 51 51](https://cloud.githubusercontent.com/assets/128547/17087456/f9a944c4-520a-11e6-8d5c-7b6b4ea51001.jpg)

**Comment by angel0d** [comment]
sorry for the late reply

@pricejn2 I have tried all you suggestions one by one and in several combinations but none seem to have worked for us

we are still having serious performance issues when massively creating/updating/deleting content

we have mitigated the problem temporarily by running specific crons only by night

this is because the most problematic cron job is from the feeds module which write/updates/deletes nodes continuously.

on New Relic we mark all cron jobs as non web transactions so on the main view of web transactions we can see normal users getting crazy high average response times when feeds is running

what we see server side (htop) is redis-server freaking out for a few seconds/minutes and maxing out to 100%

I 've been using monitor in redis-cli to catch some strange/huge happening during a redis freak out but nothing

only thing that seemed to have helped in our case was remove locking from redis via the boa site ini file with

`redis_lock_enable = FALSE`

before we would get a blocking redis::lockWait of several seconds in the traces of new relic that happened during the redis was using 100% of CPU

and this is our experience regarding the spikes and the average response times jumping to the roof because of very slow blocking redis behavior

===

regarding the latency on the other hand we found out that using the redis module (under /data/all/... etc) shipped with BOA 2.4 series is bringing average response times from 3-10ms  down to 0.3-3ms 

===

all the above are happening on a linode VM of 16 cores and a load of 2 to 5

===

only thing we haven't tried is downgrade redis itself to 2.8.17 which is still the version debian 8 ships with and the version also used bye BOA 2.4

we tried on our dev server downgrading to redis-server 2.8.17 via an `apt-get install redis-server` and keeping all the conf and seemed to have worked fine

also upgrading again to redis 3.2.1 by running a barracuda up-stable seemed to also have worked fine

@omega8cc is there any caveat in this process that could make things go really bad that I am not aware of ?

**Comment by omega8cc** [comment]
@angel0d I think that it could be a good idea to report your experience on https://www.drupal.org/project/redis

It seems that the key here is: 

> when massively creating/updating/deleting content

This is probably not too often used method to manage content, so maybe the problem goes unnoticed, if it is really not site specific.

That said, it should be helpful to know that the old Redis version doesn't suffer from it.

**Comment by pricejn2** [comment]
@angel0d do you have reduced (or erratic) New Relic reported PHP performance with the legacy Redis module? After some quick testing on a high-traffic site, I also see decreased (New Relic reported) Redis latency when using `redis_use_modern = FALSE` but this seems to come at a cost to irregular PHP latency spikes.

Is there a cache_feeds bin that could be excluded? It would seem like most of what feeds is doing doesn't need to be going through Redis at all. 

I agree with omega8 to report this on the d.o. issues. The BOA audience is quite limited relative to all redis module users.

**Comment by angel0d** [comment]
@pricejn2 @omega8cc it has past almost a year but I now have 2 small updates/notes to offer

the 1st one is long due and has been our work around most of this time. apparently memcached does not suffer from this issue and does not have any erratic behavior

the 2nd is an approach we 've testing since a few days and so far so good. the idea came after reading the following article

http://openmymind.net/Redis-SlowLog-and-DEL-surprisingly-bad-performance/

TLDR; don't delete something from Redis - Rename it and then "specially delete it" or "garbage collect it"

In the case of BOA with its daily health scripts in place a site simply needs to make through the day as restarted by night

so I patched :
- lib/Redis/Cache/Base.php
- lib/Redis/Cache/PhpRedis.php

as shown at the end.

Not only there are no hiccups but Redis (and the site itself) performs much faster.

A specific key transaction on New Relic regarding the node page 
- that has been averaging between 1.050ms to 1.150ms on memcache
- it goes down to 950ms to 1.050 with Redis (un-patched and before any erratic behavior starts)
- it goes further down to 850ms to 950ms with the patch

off course this workaround is super needy of RAM but in our case we had several GBs that we could spare

- lib/Redis/Cache/Base.php
```
@@ -16,10 +16,12 @@ abstract class Redis_Cache_Base extends Redis_AbstractBackend
     /**
      * Delete by prefix lua script
      */
+
     const EVAL_DELETE_PREFIX = <<<EOT
 local keys = redis.call("KEYS", ARGV[1])
 for i, k in ipairs(keys) do
-    redis.call("DEL", k)
+    local kk = k .. os.clock()
+    redis.("RENAME", k, kk)
 end
 return 1
 EOT;
@@ -27,11 +29,13 @@ EOT;
     /**
      * Delete volatile by prefix lua script
      */
+
     const EVAL_DELETE_VOLATILE = <<<EOT
 local keys = redis.call('KEYS', ARGV[1])
 for i, k in ipairs(keys) do
     if "1" == redis.call("HGET", k, "volatile") then
-        redis.call("DEL", k)
+        local kk = k .. os.clock()
+        redis.("RENAME", k, kk)
     end
 end
 return 1
```

- lib/Redis/Cache/PhpRedis.php

```
@@ -105,16 +105,21 @@ class Redis_Cache_PhpRedis extends Redis_Cache_Base
 
     public function delete($id)
     {
-        $this->getClient()->del($this->getKey($id));
+        $t = microtime(true);
+        $k = $this->getKey($id);
+        $this->getClient()->rename($k, $k . ':' . $t);
     }
 
     public function deleteMultiple(array $idList)
     {
         $client = $this->getClient();
-
+        $t = microtime(true);
         $pipe = $client->multi(Redis::PIPELINE);
         foreach ($idList as $id) {
-            $pipe->del($this->getKey($id));
+
+           $k = $this->getKey($id);
+           $pipe->rename($k, $k . ':' . $t);
+
         }
         // Don't care if something failed.
         $pipe->exec();

```

**Comment by omega8cc** [comment]
Interesting trick.

Also worth further testing is this patch (we were not impressed initially, though..) 

[New feature: option to use SCAN command instead of KEYS on cache wildcard deletions](https://www.drupal.org/node/2851625)

**Comment by angel0d** [comment]
I wasn't aware of the issue so I wondered why lib/Redis/Cache/Base.php was different between between production (up to date) and stage (not so up to date)

I 'll give it a try also in the context of our problem

**Comment by omega8cc** [comment]
That's really old thread but a good starting point to explore possible improvements in the Redis config and even patches.

If anyone has tested further improvements, please share!

**Comment by angel0d** [comment]
to my experience all these years the only important setting are the lazy settings that could/should all be yes

AFAIK this commit https://github.com/omega8cc/boa/commit/e160f4fce0232c941db598ccd9f04c967cc665a2 is safe to use and can help avoid most of the redis performance issues on delete/expire/eviction operations

I am really glad to see this changed

(and I can finally remove my shameful cron :| )

```
root@web:~# cat /etc/cron.hourly/fix-redis                                                                                              
#!/bin/bash

# re-configure permanent configuration in file.
sed -i "s/^maxmemory [0-9]*MB/maxmemory 32000MB/g" /etc/redis/redis.conf
sed -i "s/^lazyfree-lazy-eviction no/lazyfree-lazy-eviction yes/g" /etc/redis/redis.conf
sed -i "s/^lazyfree-lazy-expire no/lazyfree-lazy-expire yes/g" /etc/redis/redis.conf
sed -i "s/^lazyfree-lazy-server-del no/lazyfree-lazy-server-del yes/g" /etc/redis/redis.conf

# re-configure currently running redis instance.
redis-cli -a $(cat /root/.redis.pass.txt) config set maxmemory                  32000MB
redis-cli -a $(cat /root/.redis.pass.txt) config set lazyfree-lazy-eviction     yes
redis-cli -a $(cat /root/.redis.pass.txt) config set lazyfree-lazy-expire       yes
redis-cli -a $(cat /root/.redis.pass.txt) config set lazyfree-lazy-server-del   yes

```

**Comment by omega8cc** [comment]
Thank you @angel0d for the feedback and helpful hints! Let's close this now. We can always reopen in the future if needed.
